Library/OcAppleImg4Lib/OcAppleImg4Lib.c

```diff
@@ -0,0 +1,316 @@
+  ASSERT (ModulusSize > 0);
+  ASSERT (Signature != NULL);
+  ASSERT (SignatureSize > 0);
+
+  if (DEROidCompare (AlgoOid, &oidSha512Rsa)) {
+    AlgoType = RSA_ALGO_TYPE_SHA512;
+  } else if (DEROidCompare (AlgoOid, &oidSha384Rsa)) {
+    AlgoType = RSA_ALGO_TYPE_SHA384;
+  } else if (DEROidCompare (AlgoOid, &oidSha256Rsa)) {
+    AlgoType = RSA_ALGO_TYPE_SHA256;
+  } else {
+    return false;
+  }
+
+  return VerifySignatureFromData (
+           Modulus,
+           ModulusSize,
+           Exponent,
+           Signature,
+           SignatureSize,
+           Data,
+           DataSize,
+           AlgoType
+           );
+}
+
+VOID
+InternalRetrieveHwInfo (
+  OUT DERImg4Environment  *Environment
+  )
+{
+  EFI_STATUS Status;
+  UINTN      DataSize;
+  //
+  // TODO: We probably want an internal structure to collect and store these.
+  //
+  ASSERT (Environment != NULL);
+
+  ZeroMem (Environment, sizeof (*Environment));
+
+  DataSize = sizeof (Environment->ecid);
+  Status = gRT->GetVariable (
+                  L"ApECID",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->ecid
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->ecid)) {
+    Environment->ecid = 0;
+  }
+
+  DataSize = sizeof (Environment->chipId);
+  Status = gRT->GetVariable (
+                  L"ApChipID",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->chipId
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->chipId)) {
+    Environment->chipId = 0;
+  }
+
+  DataSize = sizeof (Environment->boardId);
+  Status = gRT->GetVariable (
+                  L"ApBoardID",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->boardId
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->boardId)) {
+    Environment->boardId = 0;
+  }
+
+  Environment->certificateEpoch = 2;
+
+  DataSize = sizeof (Environment->securityDomain);
+  Status = gRT->GetVariable (
+                  L"ApSecurityDomain",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->securityDomain
+                  );
+  if (EFI_ERROR (Status) || DataSize != sizeof (Environment->securityDomain)) {
+    Environment->securityDomain = 1;
+  }
+
+  DataSize = sizeof (Environment->productionStatus);
+  Status = gRT->GetVariable (
+                  L"ApProductionStatus",
+                  &gAppleSecureBootVariableGuid,
+                  NULL,
+                  &DataSize,
+                  &Environment->productionStatus
+                  );
+  if (EFI_ERROR (Status)
+   || DataSize != sizeof (Environment->productionStatus)) {
+    Environment->productionStatus = TRUE;
```

Comment:
It could be added for self-documentation reasons, but the default is TRUE, so if it's any value but TRUE, it'll yield TRUE in all boolean usages anyway. Maybe a comment?

Reply:
Replied above, it applies to all such cases.